common/lockmanager_test.go

```diff
@@ -0,0 +1,355 @@
+
+    Lock(key1);
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+}
+
+// Lock(1) -> Unlock(1) -> Unlock(1).
+func TestUnlockingAnUnlockedLock(test *testing.T) {
+    Clear();
+    defer Clear();
+
+    key1 := "testkey1";
+
+    Lock(key1);
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+
+    err = Unlock(key1);
+    if (err == nil) {
+        test.Fatalf("Failed to return an error when unlocking a lock that is already unlocked.");
+    }
+}
+
+// Unlock(1).
+func TestUnlockingKeyThatDoesntExist(test *testing.T) {
+    Clear();
+    defer Clear();
+
+    doesNotExistKey := "dne";
+
+    err := Unlock(doesNotExistKey);
+    if (err == nil) {
+        test.Fatalf("Failed to return an error when unlocking a key that does not exist.");
+    }
+}
+
+// Lock(1) -> %Lock(2) -> Unlock(1) -> %Unlock(2).
+// % denotes a thread.
+func TestThread1UnlockFirst(test *testing.T) {
+    Clear();
+    defer Clear();
+
+    key1 := "testkey1";
+    key2 := "testkey2";
+    var key1UnlockBlock sync.WaitGroup;
+    var key2LockedBlock sync.WaitGroup;
+    var threadFinishBlock sync.WaitGroup;
+
+    Lock(key1);
+
+    threadFinishBlock.Add(1);
+    key2LockedBlock.Add(1);
+    key1UnlockBlock.Add(1);
+
+    go func() {
+        defer threadFinishBlock.Done();
+
+        // Lock key2 while key1 is already locked.
+        Lock(key2);
+
+        key2LockedBlock.Done();
+
+        // Wait for the main thread to unlock key1.
+        key1UnlockBlock.Wait();
+
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock inside the thread.");
+        }
+    }()
+
+    // Wait for the thread to lock key2 before unlocking key1.
+    key2LockedBlock.Wait();
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+
+    key1UnlockBlock.Done();
+
+    // Wait for the thread to finish unlocking key2.
+    threadFinishBlock.Wait();
+}
+
+// Lock(1) -> %Lock(2) -> %Unlock(2) -> Unlock(1).
+// % denotes a thread.
+func TestThread2UnlockFirst(test *testing.T) {
+    Clear();
+    defer Clear();
+
+    key1 := "testkey1";
+    key2 := "testkey2";
+    var key2LockUnlockBlock sync.WaitGroup;
+
+    Lock(key1);
+
+    key2LockUnlockBlock.Add(1);
+
+    go func() {
+        defer key2LockUnlockBlock.Done();
+
+        // Lock key2 while key1 is already locked.
+        Lock(key2);
+
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock inside the thread.");
+        }
+    }()
+
+    // Wait for the thread to lock and unlock key2 before unlocking key1.
+    key2LockUnlockBlock.Wait();
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+}
+
+// Lock(1) -> %Lock(1) -> Unlock(1) -> %Unlock(1).
+// % denotes a thread.
+func TestLockingTwiceWithSameKeyConcurrently(test *testing.T) {
```

Comment:
Currently, there is nothing in the test that makes sure that this didn't actually happen:
Lock(1) -> %Lock(1) -> %Unlock(1) -> Unlock(1).

If the code is working correctly this won't happen, but the whole point of tests is to ensure that we are working correctly.
So, you need a check before the first Unlock to make sure that the child thread has not locked yet.
I can see a couple ways to do this (and there are probably more):
 - Check the lock timestamp after the first lock and ensure that is has not changed before the first unlock.
 - Write to a channel after the lock in the child thread. Make sure that the channel is empty in the main thread before its unlock.